doc/bu-xthin-protocol.md

```diff
@@ -0,0 +1,382 @@
+ 
+An xthinblock contains the block header followed by all the transaction hashes in the block, but only the first 64 bits of the hash is included. Finally
+after checking the block against the bloom filter provided in the GET_XTHIN message, we attach any missing transactions that we believe the requesting
+peer needs.
+
+NOTE:  You must keep the order of the hashes identical to what is in the block!
+NOTE:  The first transaction in the block, the `coinbase`, must always be added to vMissingTx.
+
+Construction of `XTHINBLOCK` is as follows:
+
+    CBlockHeader header;                        // block header
+    std::vector<uint64_t> vTxHashes;            // vector of all 64 bit transactions id's in the block
+    std::vector<CTransaction> vMissingTx;       // vector of transactions that did not match the bloom filter and the requesting peer needs
+
+
+
+
+### **NetMsgType::THINBLOCK**
+ 
+A thinblock contains the block header followed by all the full transaction hashes in the block. Finally after checking the 
+block against the bloom filter provided in the GET_XTHIN message, we attach any missing transactions that we believe the requesting
+peer needs.
+
+NOTE:  You must keep the order of the hashes identical to what is in the block!
+NOTE:  The first transaction in the block, the `coinbase`, must always be added to vMissingTx.
+
+Construction of `THINBLOCK` is as follows:
+
+    CBlockHeader header;                        // block header
+    std::vector<uint256> vTxHashes;             // vector of all transactions id's in the block
+    std::vector<CTransaction> vMissingTx;       // vector of transactions that did not match the bloom filter and the requesting peer needs
+
+
+
+
+
+### **NetMsgType::GET_XBLOCKTX**
+  
+This message is used to re-request any missing transactions that were not in the mempool but are still needed in order to reconstruct
+the block.
+
+Construction of `GET_XBLOCKTX` is as follows:
+
+    uint256 blockhash;                          // the block hash
+    std::set<uint64_t> setCheapHashesToRequest; // 64bit transactions hashes to be requested that were missing from the mempool
+
+
+
+
+### **NetMsgType::XBLOCKTX**
+ 
+This message is used to send back the transactions that were re-requested and missing from the requesting peer's memory pool
+
+Construction of `XBLOCKTX` is as follows:
+
+    uint256 blockhash;                          // the block hash
+    std::vector<CTransaction> vMissingTx;       // missing transactions to be returned
+
+
+
+
+Order and logic of P2P message requests
+---------------------------------------
+
+- `NODE A:`  receives block announcement from NODE B by an INV message and then creates and issues a GET_XTHIN to NODE B
+
+- `NODE B:`  receives GET_XTHIN and loads the attached bloom filter and creates an XTHINBLOCK which is sent back to NODE A
+         (the xthinblock is checked before sending back to see if a hash collision is within the block, if so then send a THINBLOCK instead)
```

Comment:
Is checking for hash collision required? Is it protocol violation to send a block with a hash collision?

It should be OK to send a full block also, not only THINBLOCK